hat/app/org/hatdex/hat/utils/AdjudicatorRequest.scala

```diff
@@ -50,49 +56,73 @@ object AdjudicatorRequestTypes {
+  //
+  private val secret = JwtSecretKey(sharedSecret)
+  val hatClaimBuiler: JwtClaimBuilder[HatClaim] = HatClaim.builder
 
-    val url =
-      s"${adjudicatorEndpoint}/v1/contracts/${contractId}/hat/${hatName}/${keyId}"
-    val eventuallyResponse = makeRequest(url, ws).get()
+  // Create HatClaim
+  private def createHatClaim(hatNameAsStr: String, contractId: ContractId): Option[HatClaim] = {
+    for {
+      hatName <- refineV[NonEmpty](hatNameAsStr).toOption
+      hatClaim = HatClaim(HatName(hatName), contractId)
+    } yield hatClaim
+  }
 
-    eventuallyResponse.map { response =>
-      response.status match {
-        case OK => {
-          val maybeArrayByte = Json.parse(response.body).asOpt[Array[Byte]]
-          maybeArrayByte match {
-            case None =>
-              Left(PublicKeyRequestFailure.InvalidPublicKeyFailure(
-                "The response was not able to be decoded into an Array[Byte]."))
-            case Some(arrayByte) => Right(PublicKeyReceived(arrayByte))
-          }
-        }
-        case _ => {
-          Left(PublicKeyRequestFailure.ServiceRespondedWithFailure(
-            s"The Adjudicator Service responded with an error: ${response.statusText}"))
-        }
+  private def createHatClaim(maybeHatName: Option[HatName], contractId: ContractId): Option[HatClaim] = {
+    for {
+      hatName <- maybeHatName
+      hatClaim = HatClaim(hatName, contractId)
+    } yield hatClaim
+  }
+
+  // Internal calls
+  def getPublicKey(hatName: HatName, contractId: ContractId, keyId: String)(implicit ec: ExecutionContext): Future[Either[PublicKeyRequestFailure, PublicKeyReceived]] = {
+    val claim = createHatClaim(Some(hatName), contractId)
+
+    val url = s"${adjudicatorEndpoint}/v1/contracts/hat/kid/${keyId}"
+
+    makeRequest(url, claim, ws) match {
+      case Some(req) => {
+        runPublicKeyRequest(req)
       }
-    } recover {
-      case e => {
-        Left(PublicKeyRequestFailure.ServiceRespondedWithFailure(
-          s"The Adjudicator Service responded with an error: ${e.getMessage}"))
+      case None => Future.successful(Left(PublicKeyRequestFailure.ServiceRespondedWithFailure(
+        s"The Adjudicator Service responded with an error.")))
+    }
+  }
+
+  def joinContract(hatName: String, contractId: ContractId)(implicit ec: ExecutionContext): Future[Either[JoinContractRequestFailure.ServiceRespondedWithFailure, ContractJoined]] = {
+    val claim = createHatClaim(hatName, contractId)
+
+    val url = s"${adjudicatorEndpoint}/v1/contracts/hat"
+
+    makeRequest(url, claim, ws) match {
+      case Some(req) => {
+        runJoinContractRequest(req, contractId)
       }
+      case None => Future.successful(Left(JoinContractRequestFailure.ServiceRespondedWithFailure(
+        s"The Adjudicator Service responded with an error.")))
```

Comment:
DRY this pattern inc error is repeated everywhere could it potentially be extracted ?